File: samples/src/commonMain/kotlin/sample/_exception/ExceptionBuilder.ts
import { runTest } from "../shared_ts/RunTest"
import { assert, assertQuiet, assertEquals, assertEqualsQuiet } from "../shared_ts/Assert"
import { sample, deezer } from '@kustom/Samples'

function assertException(e: deezer.kustomexport.Exception, msg: string, ...klass: any[]) {
    var name = e.constructor.name
    if (name.endsWith("_0")) {
        // Remove the '_0' added by Kotlin when using user types
        name = name.substring(0, name.length - 2)
    }

    klass.forEach(function (k) {
        assertQuiet(e instanceof k, name + " instanceof " + k.name)
    })

    assertEqualsQuiet(e.message, msg, name + " has original message")
    var consumer = new sample._exception.js.ExceptionConsumer()
    assert(consumer.consume(e).startsWith(name + "=" + msg), name + " type is preserved")
}

runTest("Exceptions", () : void => {
    var consumer = new sample._exception.js.ExceptionConsumer()
   var myEx1 = new sample._exception.js.MyEx1()// msg="hello" in Kotlin
    assertException(myEx1, "hello",
        sample._exception.js.MyEx1, Error, deezer.kustomexport.Exception,
        deezer.kustomexport.RuntimeException, deezer.kustomexport.IllegalStateException)

    var myEx2 = new sample._exception.js.MyEx2("testing MyEx2")
    assertException(myEx2, "testing MyEx2",
        sample._exception.js.MyEx2, Error, deezer.kustomexport.Exception,
        deezer.kustomexport.RuntimeException, deezer.kustomexport.IllegalStateException)
    assertEqualsQuiet("custom exception val", myEx2.ex2Bonus, "custom field")

    var builder = new sample._exception.js.ExceptionBuilder()
    //assertException(builder.buildError("error"), "error", deezer.kustomexport.Error)
    //assertException(builder.buildException("exception"), "exception", deezer.kustomexport.Exception)
    assertException(builder.buildRuntimeException("re"), "re",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException)
    assertException(builder.buildIllegalArgumentException("iae"), "iae",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.IllegalArgumentException)
    assertException(builder.buildIllegalStateException("ise"), "ise",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.IllegalStateException)
    assertException(builder.buildIndexOutOfBoundsException("ioobe"), "ioobe",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.IndexOutOfBoundsException)
    assertException(builder.buildConcurrentModificationException("cme"), "cme",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.ConcurrentModificationException)
    assertException(builder.buildUnsupportedOperationException("uoe"), "uoe",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.UnsupportedOperationException)
    assertException(builder.buildNumberFormatException("nfe"), "nfe",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.IllegalArgumentException, deezer.kustomexport.NumberFormatException)
    assertException(builder.buildNullPointerException("uoe"), "uoe",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.NullPointerException)
    assertException(builder.buildClassCastException("cce"), "cce",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.ClassCastException)
    //assertException(builder.buildAssertionError("az"), "az",
        //deezer.kustomexport.AssertionError, deezer.kustomexport.Error)
    assertException(builder.buildNoSuchElementException("nsee"), "nsee",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.NoSuchElementException)
    assertException(builder.buildArithmeticException("ae"), "ae",
        Error, deezer.kustomexport.Exception, deezer.kustomexport.RuntimeException, deezer.kustomexport.ArithmeticException)
})

File: samples/src/commonMain/kotlin/sample/_exception/ExceptionBuilder.kt
package sample._exception

import deezer.kustomexport.KustomExport
import kotlinx.coroutines.CancellationException

@KustomExport
class ExceptionBuilder {
    fun buildArithmeticException(msg: String) = ArithmeticException(msg)

    //fun buildAssertionError(msg: String) = AssertionError(msg)
    fun buildClassCastException(msg: String) = ClassCastException(msg)
    fun buildConcurrentModificationException(msg: String) = ConcurrentModificationException(msg)

    //fun buildError(msg: String) = Error(msg)
    fun buildException(msg: String) = Exception(msg)
    fun buildIllegalArgumentException(msg: String) = IllegalArgumentException(msg)
    fun buildIllegalStateException(msg: String) = IllegalStateException(msg)
    fun buildIndexOutOfBoundsException(msg: String) = IndexOutOfBoundsException(msg)
    fun buildNoSuchElementException(msg: String) = NoSuchElementException(msg)
    fun buildNullPointerException(msg: String) = NullPointerException(msg)
    fun buildNumberFormatException(msg: String) = NumberFormatException(msg)
    fun buildRuntimeException(msg: String) = RuntimeException(msg)
    fun buildUnsupportedOperationException(msg: String) = UnsupportedOperationException(msg)

    // Coroutines
    fun buildCancellationException(msg: String) = CancellationException(msg)
}

@KustomExport
class MyEx1 : IllegalStateException("hello")

@KustomExport
class MyEx2(msg: String) : IllegalStateException(msg) {
    val ex2Bonus = "custom exception val"
}


@KustomExport
class ExceptionConsumer {
    fun consume(e: Exception): String {
        @Suppress("USELESS_IS_CHECK") // Cause typescript can go crazy
        return when (e) {
            is CancellationException -> "CancellationException=${e.message}"

            is MyEx1 -> "MyEx1=${e.message}"
            is MyEx2 -> "MyEx2=${e.message}"
            is ArithmeticException -> "ArithmeticException=${e.message}"
            is ClassCastException -> "ClassCastException=${e.message}"
            is ConcurrentModificationException -> "ConcurrentModificationException=${e.message}"
            is NumberFormatException -> "NumberFormatException=${e.message}"
            is IllegalStateException -> "IllegalStateException=${e.message}"
            is IndexOutOfBoundsException -> "IndexOutOfBoundsException=${e.message}"
            is NoSuchElementException -> "NoSuchElementException=${e.message}"
            is NullPointerException -> "NullPointerException=${e.message}"
            is UnsupportedOperationException -> "UnsupportedOperationException=${e.message}"

            is IllegalArgumentException -> "IllegalArgumentException=${e.message}"

            is RuntimeException -> "RuntimeException=${e.message}"

            // TODO :
            //is AssertionError -> "AssertionError=${e.message}"
            //is Error -> "Error=${e.message}"
            is Exception -> "Exception=${e.message}"

            else -> "Not an exception!!!"
        }
    }
}

// Can Exception subclasses
@KustomExport
class FooException(msg: String, val e: Exception) : Exception(msg)

// Can other exception type subclasses
@KustomExport
class MyISException(msg: String) : IllegalStateException(msg)

@KustomExport
sealed class SealedException(msg: String) : Exception(msg)

@KustomExport
class FirstSealedException : SealedException("first")
